x/ecocredit/server/basket/msg_put_test.go

```diff
@@ -0,0 +1,478 @@
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.alice.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: s.batchDenom,
+				Amount:     a,
+			},
+		},
+	})
+}
+
+func (s *putSuite) AliceAttemptsToPutCreditsFromCreditBatchIntoTheBasket(a string) {
+	amount, _ := sdk.NewIntFromString(s.tradableCredits)
+	coins := sdk.NewCoins(sdk.NewCoin(s.basketDenom, amount))
+
+	s.bankKeeper.EXPECT().
+		MintCoins(s.sdkCtx, basket.BasketSubModuleName, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	s.bankKeeper.EXPECT().
+		SendCoinsFromModuleToAccount(s.sdkCtx, basket.BasketSubModuleName, s.alice, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.alice.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: a,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) BobAttemptsToPutCreditsFromCreditBatchIntoTheBasket(a string) {
+	amount, _ := sdk.NewIntFromString(s.tradableCredits)
+	coins := sdk.NewCoins(sdk.NewCoin(s.basketDenom, amount))
+
+	s.bankKeeper.EXPECT().
+		MintCoins(s.sdkCtx, basket.BasketSubModuleName, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	s.bankKeeper.EXPECT().
+		SendCoinsFromModuleToAccount(s.sdkCtx, basket.BasketSubModuleName, s.bob, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.bob.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: a,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) AliceAttemptsToPutCreditsIntoTheBasket() {
+	amount, _ := sdk.NewIntFromString(s.tradableCredits)
+	coins := sdk.NewCoins(sdk.NewCoin(s.basketDenom, amount))
+
+	s.bankKeeper.EXPECT().
+		MintCoins(s.sdkCtx, basket.BasketSubModuleName, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	s.bankKeeper.EXPECT().
+		SendCoinsFromModuleToAccount(s.sdkCtx, basket.BasketSubModuleName, s.alice, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.alice.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: s.batchDenom,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) TheBasketHasACreditBalanceWithAmount(a string) {
+	basket, err := s.stateStore.BasketTable().GetByBasketDenom(s.ctx, s.basketDenom)
+	require.NoError(s.t, err)
+
+	balance, err := s.stateStore.BasketBalanceTable().Get(s.ctx, basket.Id, s.batchDenom)
+	require.NoError(s.t, err)
+
+	require.Equal(s.t, a, balance.Balance)
+}
+
+func (s *putSuite) TheBasketTokenHasATotalSupplyWithAmount(a string) {
```

Comment:
Not exactly sure what you mean. This is a step used to check the scenarios in which we are testing an amount of credits are put into the basket, i.e. "Rule: The user must have a credit balance more than or equal to the credits being put into the basket". Each of these functions are steps combined to complete a test scenario. We could potentially break this out into multiple scenarios each checking one of the following:

```
      Then the basket has a credit balance with amount "<put-amount>"
      And the basket token has a total supply with amount "<put-amount>"
      And alice has a credit balance with amount "<user-balance-after>"
      And alice has a basket token balance with amount "<put-amount>"
```

If we did that, there would still be no failing scenario. We could add a scenario where credits are put into the basket twice and we expect the amount of basket tokens minted to be equal to the sum of the equivalent amount of credits put into the basket but there would still be no failing scenario unless we manually added tokens to the supply but then we would no longer be testing the behavior of the put method and merely writing a test for the sake of a failing scenario.

Reply:
Not exactly sure what you mean. This is a step used to check the scenarios in which we are testing an amount of credits are put into the basket, i.e. "Rule: The user must have a credit balance more than or equal to the credits being put into the basket". Each of these functions are steps combined to complete a test scenario. We could potentially break this out into multiple scenarios each checking one of the following:

```
      Then the basket has a credit balance with amount "<put-amount>"
      And the basket token has a total supply with amount "<put-amount>"
      And alice has a credit balance with amount "<user-balance-after>"
      And alice has a basket token balance with amount "<put-amount>"
```

If we did that, there would still be no failing scenario. We could add a scenario where credits are put into the basket twice and we expect the amount of basket tokens minted to be equal to the sum of the equivalent amount of credits put into the basket but there would still be no failing scenario unless we manually added tokens to the supply but then we would no longer be testing the behavior of the put method and writing a test for the sake of a failing scenario.

Reply:
The scenario I'm concerned with is `Scenario Outline: basket token supply is updated`. Are token _supplies_ really our concern? since bank is mocked and holds no state, I don't really see the value in it. Wouldn't the behavior of token supplies being updated be better slated for an integration test with a real bank module?

Reply:
I think we want to test as much as possible in the unit tests and not be dependent on integration tests. It should probably be in both, here in the unit test and in the integration test.

Reply:
but it's not actually testing anything if it's mocked. the function above is equivalent to:

```go
	amount, err := strconv.ParseInt(a, 10, 32)
	require.NoError(s.t, err)
	coin := sdk.NewInt64Coin(s.basketDenom, amount)
       supply := coin
       require.Equal(s.t, coin, supply)
```

Reply:
Ok, I see what you are saying although I do think it's valuable to capture this in what is essentially a specification for the intended behavior of the put message. What are we expecting when the put message is executed? We are expecting the user balance to be updated, the basket credit balance to be updated, the user token balance to be updated, and the basket token supply to be updated. The primary purpose of acceptance tests is to document behavior. As far as I understand, this is aligned with the goal of writing feature files and acceptance tests.

Reply:
if its for spec/doc purposes, ok 👍🏻 was just unsure how/if mocked units fit into BDD. I can resolve this and the other comment.

Reply:
If it does no harm, I think it would we should keep it for now and we can revisit later. You might be right but I think this still holds some value as a scenario. Would be happy to have you dig into writing BDD style acceptance tests and work with me on how we can improve the structure. A couple books that I've read that might be of interest:

https://leanpub.com/bddbooks-discovery
https://leanpub.com/bddbooks-formulation

If you are interested, we can purchase copies for you as well.

Reply:
I should also add that this has been very challenging and appreciate the questioning because I'm not sure I have the right answer necessarily. Definitely a work in progress.

Reply:
yeah id be interested in giving those a read!

I will add as well this PR is very informative and well put together, I feel pretty confident I could try my hand at implementing some acceptance tests using this as a base/example. Good stuff! 🦾